src/main/java/org/openjdk/jextract/impl/ClassSourceBuilder.java

```diff
@@ -50,160 +47,104 @@ enum Kind {
+        // for a (nested) class 'com.foo.package.A.B.C' this will return 'C'
+        return className;
     }
 
-    String className() {
-        return desc.displayName();
+    final String fullName() {
+        // for a (nested) class 'com.foo.package.A.B.C' this will return 'A.B.C'
+        return Stream.concat(enclosingClassNames.stream(), Stream.of(className)).collect(Collectors.joining("."));
     }
 
-    String fullName() {
-        return isNested() ?
-                ((ClassSourceBuilder)enclosing).className() + "." + className() :
-                className();
+    final SourceFileBuilder sourceFileBuilder() {
+        return sb;
     }
 
-    @Override
-    public final String packageName() {
-        return desc.packageName();
+    final List<String> enclosingClassNames() {
+        return enclosingClassNames;
     }
 
-    String superClass() {
-        return null;
-    }
-
-    String mods() {
-        if (kind == Kind.INTERFACE) {
-            return "public ";
-        }
-        return (isNested() ? "public static " : "public ") +
-                (isClassFinal() ? "final " : "");
-    }
-
-    boolean isClassFinal() {
-        return true;
-    }
-
-    void classBegin() {
-        if (isNested()) {
-            incrAlign();
-        }
-        emitPackagePrefix();
-        emitImportSection();
-
-        classDeclBegin();
+    final void classBegin() {
         indent();
-        append(mods());
-        append(kind.kindName + " " + className());
-        if (superClass() != null) {
+        append(modifiers);
+        append(" ");
+        append(kind.kindName + " " + className);
+        if (superName != null) {
             append(" extends ");
-            append(superClass());
+            append(superName);
         }
         append(" {\n\n");
-        if (kind != Kind.INTERFACE) {
-            emitConstructor();
-        }
     }
 
-    void classDeclBegin() {}
-
-    void emitConstructor() {
-        incrAlign();
-        indent();
-        append("// Suppresses default constructor, ensuring non-instantiability.\n");
-        indent();
-        append("private ");
-        append(className());
-        append("() {}");
-        append('\n');
-        decrAlign();
-    }
-
-    JavaSourceBuilder classEnd() {
+    final void classEnd() {
         indent();
         append("}\n\n");
-        if (isNested()) {
-            decrAlign();
-            ((ClassSourceBuilder)enclosing).append(build());
-            sb = null;
-        }
-        return enclosing;
-    }
-
-    @Override
-    public List<JavaFileObject> toFiles() {
-        if (isNested()) {
-            throw new UnsupportedOperationException("Nested builder!");
-        }
-        String res = build();
-        sb = null;
-        return List.of(Utils.fileFromString(packageName(), className(), res));
     }
 
     // Internal generation helpers (used by other builders)
 
-    void append(Object o) {
-        sb.append(o);
-    }
```

Comment:
This overload, and the one accepting a `boolean` were unused, so I've dropped them.